File: firebase/functions/lib/events/live_meetings/toggle_like_dislike_on_meeting_user_suggestion.dart
import 'dart:async';

import 'package:firebase_admin_interop/firebase_admin_interop.dart';
import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import '../../on_call_function.dart';
import '../../utils/infra/firestore_utils.dart';
import 'package:data_models/discussion_threads/discussion_thread.dart';
import 'package:data_models/events/live_meetings/meeting_guide.dart';

class ToggleLikeDislikeOnMeetingUserSuggestion
    extends OnCallMethod<ParticipantAgendaItemDetailsMeta> {
  ToggleLikeDislikeOnMeetingUserSuggestion()
      : super(
          'toggleLikeDislikeOnMeetingUserSuggestion',
          (jsonMap) => ParticipantAgendaItemDetailsMeta.fromJson(jsonMap),
        );

  @override
  Future<void> action(
    ParticipantAgendaItemDetailsMeta request,
    CallableContext context,
  ) async {
    await firestore.runTransaction((transaction) async {
      final agendaItemDetailsDocRef = firestore.document(request.documentPath);
      final agendaItemDetailsSnap =
          await transaction.get(agendaItemDetailsDocRef);
      if (!agendaItemDetailsSnap.exists) {
        print(
          'agendaItemDetailsSnap from path ${request.documentPath} does not exist',
        );
        return;
      }
      final ParticipantAgendaItemDetails agendaItemDetails;
      final agendaItemDetailsDataMap = agendaItemDetailsSnap.data.toMap();

      try {
        agendaItemDetails = ParticipantAgendaItemDetails.fromJson(
          firestoreUtils.fromFirestoreJson(agendaItemDetailsDataMap),
        );
      } catch (e) {
        print(
          'Cannot parse participant agenda item details: $agendaItemDetailsDataMap',
        );
        return;
      }

      final userSuggestion = agendaItemDetails.suggestions
          .where(
            (element) => element.id == request.userSuggestionId,
          )
          .firstOrNull;
      if (userSuggestion == null) {
        print('userSuggestion is null. Request: ${request.toJson()}');
        return;
      }

      final voterId = request.voterId;
      print(
        'Toggle: ${request.likeType} for user suggestion ${userSuggestion.id}',
      );

      switch (request.likeType) {
        case LikeType.like:
          if (!userSuggestion.likedByIds.any((element) => element == voterId)) {
            userSuggestion.likedByIds.add(voterId);
          }
          userSuggestion.dislikedByIds
              .removeWhere((element) => element == voterId);
          break;
        case LikeType.neutral:
          userSuggestion.likedByIds
              .removeWhere((element) => element == voterId);
          userSuggestion.dislikedByIds
              .removeWhere((element) => element == voterId);
          break;
        case LikeType.dislike:
          userSuggestion.likedByIds
              .removeWhere((element) => element == voterId);
          if (!userSuggestion.dislikedByIds
              .any((element) => element == voterId)) {
            userSuggestion.dislikedByIds.add(voterId);
          }
          break;
      }

      final updateDataMap = agendaItemDetails.toJson();
      print('Update data: $updateDataMap');

      transaction.update(
        agendaItemDetailsDocRef,
        UpdateData.fromMap(firestoreUtils.toFirestoreJson(updateDataMap)),
      );
    });
  }
}

File: firebase/functions/src/events/live_meetings/toggle_like_dislike_on_meeting_user_suggestion.ts
import * as functions from 'firebase-functions';
import * as admin from 'firebase-admin';
import { OnCallMethod } from '../../on_call_function';
import { firestore, firestoreUtils } from '../../utils/infra/firestore_utils';
import { ParticipantAgendaItemDetails } from '../../types';

interface ParticipantAgendaItemDetailsMeta {
  documentPath: string;
  userSuggestionId: string;
  voterId: string;
  likeType: 'like' | 'neutral' | 'dislike';
}

export class ToggleLikeDislikeOnMeetingUserSuggestion extends OnCallMethod<ParticipantAgendaItemDetailsMeta> {
  constructor() {
    super(
      'toggleLikeDislikeOnMeetingUserSuggestion',
      (jsonMap) => jsonMap as ParticipantAgendaItemDetailsMeta
    );
  }

  async action(request: ParticipantAgendaItemDetailsMeta, _context: functions.https.CallableContext): Promise<void> {
    await firestore.runTransaction(async (transaction) => {
      const docRef = firestore.doc(request.documentPath);
      const snap = await transaction.get(docRef);
      if (!snap.exists) {
        console.log(`agendaItemDetailsSnap from path ${request.documentPath} does not exist`);
        return;
      }

      let agendaItemDetails: ParticipantAgendaItemDetails;
      try {
        agendaItemDetails = firestoreUtils.fromFirestoreJson(snap.data() ?? {}) as unknown as ParticipantAgendaItemDetails;
      } catch (e) {
        console.log('Cannot parse participant agenda item details:', snap.data());
        return;
      }

      const suggestions = ((agendaItemDetails as any).suggestions ?? []) as Array<any>;
      const suggestionIndex = suggestions.findIndex((s: any) => s.id === request.userSuggestionId);
      if (suggestionIndex < 0) {
        console.log('userSuggestion is null. Request:', request);
        return;
      }

      const suggestion = { ...(suggestions[suggestionIndex] as any) };
      const likedByIds: string[] = [...(suggestion.likedByIds ?? [])];
      const dislikedByIds: string[] = [...(suggestion.dislikedByIds ?? [])];
      const voterId = request.voterId;

      console.log(`Toggle: ${request.likeType} for user suggestion ${suggestion.id}`);

      if (request.likeType === 'like') {
        if (!likedByIds.includes(voterId)) likedByIds.push(voterId);
        const idx = dislikedByIds.indexOf(voterId);
        if (idx >= 0) dislikedByIds.splice(idx, 1);
      } else if (request.likeType === 'neutral') {
        const li = likedByIds.indexOf(voterId);
        if (li >= 0) likedByIds.splice(li, 1);
        const di = dislikedByIds.indexOf(voterId);
        if (di >= 0) dislikedByIds.splice(di, 1);
      } else if (request.likeType === 'dislike') {
        const li = likedByIds.indexOf(voterId);
        if (li >= 0) likedByIds.splice(li, 1);
        if (!dislikedByIds.includes(voterId)) dislikedByIds.push(voterId);
      }

      suggestion.likedByIds = likedByIds;
      suggestion.dislikedByIds = dislikedByIds;
      suggestions[suggestionIndex] = suggestion;

      const updateData = firestoreUtils.toFirestoreJson({ suggestions } as Record<string, unknown>);
      transaction.update(docRef, updateData);
    });
  }
}
